Document dynamic changesetTemplates and steps.outputs (#17286)
sourcegraph/sourcegraph: doc/campaigns/references/campaign_spec_templating.md

.markdown-body pre.chroma { font-size: 0.75em; }
</style>

## Overview

[Certain fields](#fields-with-template-support) in a [campaign spec YAML](campaign_spec_yaml_reference.md) support templating to create even more powerful and performant campaigns.

Templating in a campaign spec uses the delimiters `${{` and `}}`. Inside the delimiters, [template variables](#template-variables) and [template helper functions](#template-helpers-functions) may be used to produce a text value.

### Example campaign spec

Here is an excerpt of a campaign spec that uses templating:

```yaml
on:
  - repositoriesMatchingQuery: lang:go fmt.Sprintf("%d", :[v]) patterntype:structural -file:vendor

steps:
  - run: comby -in-place 'fmt.Sprintf("%d", :[v])' 'strconv.Itoa(:[v])' ${{ join repository.search_result_paths " " }}
    #                                                                   ^ templating starts here
    container: comby/comby
  - run: goimports -w ${{ join previous_step.modified_files " " }}
    #                 ^ templating starts here
    container: unibeautify/goimports
```

> Before executing the first `run` command `repository.search_result_paths` will be replaced with the relative-to-root-dir file paths of each search resulted yielded by `repositoriesMatchingQuery`. By using the [template helper function](#template-helper-functions) `join`, an argument list of whitespace-separated values is constructed.

Reviewer: ```suggestion
Before executing the first `run` command, `repository.search_result_paths` will be replaced with the relative-to-root-dir file paths of each search result yielded by `repositoriesMatchingQuery`. By using the [template helper function](#template-helper-functions) `join`, an argument list of whitespace-separated values is constructed.
```

- [`changesetTemplate.body`](campaign_spec_yaml_reference.md#changesettemplate-body)
- [`changesetTemplate.branch`](campaign_spec_yaml_reference.md#changesettemplate-branch)
- [`changesetTemplate.commit.message`](campaign_spec_yaml_reference.md#changesettemplate-commit-message)
- [`changesetTemplate.commit.author.name`](campaign_spec_yaml_reference.md#changesettemplate-commit-author)
- [`changesetTemplate.commit.author.email`](campaign_spec_yaml_reference.md#changesettemplate-commit-author)

## Template variables

Template variables are the names that are defined and accessible when using templating syntax in a given context.

Depending on the context in which templating is used, different variables are available.

For example: in the context of `steps` the template variable `previous_step` is available, but not in the context of `changesetTemplate`.

### `steps` context

The following template variables are available in the fields under `steps`.

They are evaluated before the execution of each entry in `steps`, except for the `step.*` variables, which only contain values _after_ the step has executed.

> | Template variable | Description |

Reviewer: How far do we want to get into typing here? Some of these values map to `[]string`, others to `string`, which means they'll need to be handled differently in practice. Should that be a column?

Author: Good idea! I'll add a column. I'll use `string` and `list of strings` as wording, though, since `[]string` might trip up some people.

Author: (I'll add these in a follow-up PR to make reviewing easier)

| --- | --- |
| `repository.search_result_paths` | Unique list of file paths relative to the repository root directory in which the search results of the `repositoriesMatchingQuery`s have been found. |
| `repository.name` | Full name of the repository in which the step is being executed. |
| `previous_step.modified_files` | List of files that have been modified by the previous step in `steps`. Empty list if no files have been modified. |
| `previous_step.added_files` | List of files that have been added by the previous step in `steps`. Empty list if no files have been added. |
| `previous_step.deleted_files` | List of files that have been deleted by the previous step in `steps`. Empty list if no files have been deleted. |
| `previous_step.stdout` | The complete output of the previous step on standard output. |
| `previous_step.stderr` | The complete output of the previous step on standard error. |

> | `step.modified_files` | Only in `steps.outputs`: List of files that have been modified by the just-executed step. Empty list if no files have been modified. </br><i><small>Requires Sourcegraph 3.24 and [Sourcegraph CLI](../../cli/index.md) 3.24 or later</small></i>. |

Reviewer: I wonder if `current` would be clearer here, by contrast with `previous` above:

```suggestion
| `step.modified_files` | Only in `steps.outputs`: List of files that have been modified by the current step. Empty list if no files have been modified. </br><i><small>Requires Sourcegraph 3.24 and [Sourcegraph CLI](../../cli/index.md) 3.24 or later</small></i>. |
```

Author: In that context it makes more sense, since it's clear that this is not available _before_ the step is executed. Thanks!

        value: "${{ step.stdout }}"
  - run: echo "We have access to the output here: ${{ outputs.myFriendlyMessage }}"
    container: alpine:3
    outputs:
      stepTwoOutput:
        otherMessage: "here too: ${{ outputs.myFriendlyMessage }}"

changesetTemplate:
  # [...]
  body: |
    The first step left us the following message: ${{ outputs.myFriendlyMessage }}
    The second step left this one: ${{ outputs.otherMessage }}
```

Using the [`steps.outputs.<name>.format`](campaign_spec_yaml_reference.md#steps-outputs-name-format) field, it's possible to parse the value of an output as JSON or YAML and access it as a data structure instead of just text:

```yaml
steps:
  - run: cat .goreleaser.yml
    container: alpine:3
    outputs:
      goreleaserConfig:
        value: "${{ step.stdout }}"
        # The step's output is parsed as YAML, making it accessible as a YAML
        # object in the other templating fields.
        format: yaml
      goreleaserConfigExists:
        # We can use the power of Go's text/template engine to dynamically produce complex values
        value: "exists: ${{ gt (len step.stderr) 0 }}"
        format: yaml

changesetTemplate:
  # [...]

>   # Since templating fields use Go's `text/templates` and `goreleaserConfig` was

Reviewer: ```suggestion
  # Since templating fields use Go's `text/template` and `goreleaserConfig` was
```
